go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -77,69 +77,64 @@ func newEgressFirewallRule(rawEgressFirewallRule egressfirewallapi.EgressFirewal
 	return efr, nil
 }
 
-// This function is used to sync egress firewall setup. It does three "cleanups"
+// This function is used to sync egress firewall setup. It does two cleanups
 
-// - 	Cleanup the old implementation (using LRP) in local GW mode -> new implementation (using ACLs) local GW mode
+// - 	Cleanup the old local GW mode implementation (using LRP)
 //  	For this it just deletes all LRP setup done for egress firewall
 
-// -	Cleanup the new local GW mode implementation (using ACLs on the node switch) -> shared GW mode implementation (using ACLs on the join switch)
-//  	For this it just deletes all ACL setup done for egress firewall on the node switches
-
-// -	Cleanup the old implementation (using LRP) in local GW mode -> shared GW mode implementation (using ACLs on the join switch)
-//  	For this it just deletes all LRP setup done for egress firewall
+// -	Cleanup the old shared GW mode implementation (using ACLs on the all switches with direction=from-lport)
+//  	For this it just deletes all the ACL setup done for egress firewall on all switches with direction=from-lport
 
 // NOTE: Utilize the fact that we know that all egress firewall related setup must have a priority: types.MinimumReservedEgressFirewallPriority <= priority <= types.EgressFirewallStartPriority
 func (oc *Controller) syncEgressFirewall(egressFirwalls []interface{}) {
-	if config.Gateway.Mode == config.GatewayModeShared {
-		// Mode is shared gateway mode, make sure to delete all ACLs on the node switches
-		egressFirewallACLIDs, stderr, err := util.RunOVNNbctl(
-			"--data=bare",
-			"--no-heading",
-			"--columns=_uuid",
-			"--format=table",
-			"find",
-			"acl",
-			fmt.Sprintf("priority<=%s", types.EgressFirewallStartPriority),
-			fmt.Sprintf("priority>=%s", types.MinimumReservedEgressFirewallPriority),
-		)
+	egressFirewallACLIDs, stderr, err := util.RunOVNNbctl(
+		"--data=bare",
+		"--no-heading",
+		"--columns=_uuid",
+		"--format=table",
+		"find",
+		"acl",
+		fmt.Sprintf("priority<=%s", types.EgressFirewallStartPriority),
+		fmt.Sprintf("priority>=%s", types.MinimumReservedEgressFirewallPriority),
+		fmt.Sprintf("direction=%s", types.DirectionFromLPort),
+	)
+	if err != nil {
+		klog.Errorf("Unable to list egress firewall logical router policies, cannot cleanup old stale data, stderr: %s, err: %v", stderr, err)
+		return
+	}
+	if egressFirewallACLIDs != "" {
+		nodes, err := oc.watchFactory.GetNodes()
 		if err != nil {
-			klog.Errorf("Unable to list egress firewall logical router policies, cannot cleanup old stale data, stderr: %s, err: %v", stderr, err)
+			klog.Errorf("Unable to cleanup egress firewall ACLs remaining from local gateway mode, cannot list nodes, err: %v", err)
 			return
 		}
-		if egressFirewallACLIDs != "" {
-			nodes, err := oc.watchFactory.GetNodes()
+		logicalSwitches := []string{types.OVNJoinSwitch}
+		for _, node := range nodes {
```

Comment:
the only concern I see with this commit is you are getting the logical switches to check based on the nodes in k8s. There could be some issue where, lets say there is a leftover logical switch in OVN for a node that isnt accounted for that also has the egress firewall acl on it. In that case, the code would miss cleaning the acl off of htat switch, and the ACL would hang around. Then in the create code, when it does the ACL search for the matching external ID, it would find one that exists and re-add the wrong ACL to every logical switch.

I wonder if we can have a better way here to ensure the ACL is gone. It would be great if we could just directly delete the ACL with an ovsdb call and that would remove it off of all the switches. Is that possible? Otherwise, could we get the list of logical switches the ACL applies to from OVN itself rather than k8s?

Reply:
Done. I've made so that we just list all ACL with a priority between the egress firewall max and min and with a direction of `from-lport`, and set the direction on them to `to-lport`. As such we don't need to list nodes.